libbpf-rs/src/object.rs

```diff
@@ -327,7 +327,9 @@ impl Object {
 
             // Get the map fd
             let fd = unsafe { libbpf_sys::bpf_map__fd(next_ptr) };
-            if fd < 0 {
+            // Old kernels can have uninitialized internal maps:
+            // https://github.com/libbpf/libbpf/blob/5c31bcf220f66e70f39fd141f5b0c55c6ab65e8e/src/libbpf.c#L5009-L5022
+            if fd < 0 && !unsafe { libbpf_sys::bpf_map__is_internal(next_ptr) } {
```

Comment:
`bpf_map__is_internal` check is handled by the generated skeleton. Why do we need it here ?

I am afraid the changes I made in #192 is wrong.

Reply:
I left reference to `libbpf` in the comment, it basically explains the reason.
For old kernels some data maps (internal) are not really created and their `fd` is `-1` after loading the object. So without relaxing the checks in `Object::new` it just won't load.

> I am afraid the changes I made in https://github.com/libbpf/libbpf-rs/pull/192 is wrong.

Sorry, I don't see the connection.

Reply:
In the generated skeleton, we have:

```rust
    pub fn load(mut self) -> libbpf_rs::Result<RunqslowerSkel<'a>> {
        let ret = unsafe { libbpf_sys::bpf_object__load_skeleton(self.skel_config.get()) };
        if ret != 0 {
            return Err(libbpf_rs::Error::System(-ret));
        }

        let obj = unsafe { libbpf_rs::Object::from_ptr(self.obj.take_ptr())? };

        Ok(RunqslowerSkel {
            obj,
            skel_config: self.skel_config,
            links: RunqslowerLinks::default(),
        })
    }
```

If we reach `Object::new`, all BPF maps and progs should be loaded (or skipped) successfully.

I think we can skip `fd < 0` check entirely.

Reply:
I'm a bit confused. You linked code from another crate. I was not even aware of it, nor have I used it.

> I think we can skip fd < 0 check entirely.

If you think it's ok, sure. I assumed there should be a reason for them to be there in the first place.

Reply:
The code I posted is geneated by libbpf-cargo (examples/runqslower), the primary user of `Object`. In this use-case, if we reach `Object::new`, we pass the libbpf skeleton load phase.

As I mentioned above, if we can't skip this check here, the change I maded in #192 is probably wrong.

Reply:
In the PR you've linked I see that you've removed similar checks for program fds, are you referring to that? I don't think this is necessarily wrong. To me, these checks look more like safety checks, but honestly I don't know.